pandemic51/core/streaming.py

```diff
@@ -245,35 +111,420 @@ def sample_first_frame(inpath, outpath, width=None, height=None):
+        return os.path.join(panc.STREAMS_DIR, stream_name + ".json")
+
+    @staticmethod
+    def get_stream_names():
+        '''Get names of all streams serialized on disk.
 
-    Args:
-        stream_name: the stream name
-        outdir: the output directory
-        width: an optional width to resize the image
-        height: an optional height to resize the image
+        Returns:
+             a list of stream_name strings
+        '''
+        _, matches = etau.parse_glob_pattern(Stream.stream_path("*"))
+        return [x[0] for x in matches]
+
+    @property
+    def path(self):
+        '''Path to the Stream serialized on disk'''
+        return self.stream_path(self.stream_name)
+
+    def get_live_stream_url(self):
+        '''Get the URL for streaming'''
+        raise NotImplementedError("Subclass must implement")
+
+    def download_image(self, outdir):
+        '''Downloads an image from the latest stream
+
+        Args:
+            outdir: the output directory
+
+        Returns:
+            is_new_img: `True` if the image was not already on disk
+            image_path: path the the downloaded image on disk
+            dt: datetime object of when the image was downloaded
+        '''
+        raise NotImplementedError("Subclass must implement")
+
+    def download_image_and_store(self, outdir):
+        '''Downloads an image from the latest stream, and add it to the
+        database.
+
+        If an image with the same filename was already on disk, no new entry is
+        added to the database.
+
+        Args:
+            stream_name: the stream name
+            outdir: the output directory
+
+        Returns:
+            image_path: path the the downloaded image on disk
+            dt: datetime object of when the image was downloaded
+        '''
+        is_new_img, image_path, dt = self.download_image(outdir)
+
+        if is_new_img:
+            add_stream_history(self.stream_name, dt, image_path)
+
+        return image_path, dt
+
+    @classmethod
+    def from_dict(cls, d, *args, **kwargs):
+        '''Constructs a Stream object from a JSON dictionary.
+
+        Args:
+            d: a JSON dictionary representation of a Serializable object
+
+        Returns:
+            a Stream instance
+        '''
+        downloader_cls = etau.get_class(d["type"])
+        return downloader_cls._from_dict(d)
+
+    @classmethod
+    def from_json(cls, path, *args, **kwargs):
+        '''Constructs a Stream object from a JSON file.
+
+        If `stream_name` key is missing, it is automatically populated.
+
+        Args:
+            path: the path to the JSON file on disk
+
+        Returns:
+            a Stream instance
+        '''
+        d = etas.read_json(path)
 
-    Returns:
-        image_path: path the the downloaded image on disk
-        dt: datetime object of when the image was downloaded
+        if "stream_name" not in d:
+            d["stream_name"] = os.path.splitext(os.path.basename(path))[0]
```

Comment:
Why allow `stream_name` to be absent? Seems like more confusion than utility. @ehofesmann asked about this, for example